pincer/commands/arg_types.py

```diff
@@ -0,0 +1,228 @@
+
+
+class _CommandTypeMeta(type):
+    def __getitem__(cls, args: Union[Tuple, Any]):
+        if not isinstance(args, tuple):
+            args = args,
+
+        return cls(*args)
+
+
+class CommandArg(metaclass=_CommandTypeMeta):
+    """
+    Holds all application command options
+
+    .. code-block:: python3
+
+        CommandArg[
+            # This is the type of command.
+            # Supported types are str, int, bool, float, User, Channel, and Role
+            int,
+            # The modifiers to the command go here
+            Description["Pick a number 1-10"],
+            MinValue[1],
+            MaxValue[10]
+        ]
+
+    Parameters
+    ----------
+    command_type : T
+        The type of the command
+    *args : :class:`pincer.commands.arg_types.Modifier`
+
+    """
+
+    def __init__(self, command_type, *args) -> None:
+        self.command_type = command_type
+        self.modifiers = args
+
+    def get_arg(self, arg_type: T) -> T:
+        for arg in self.modifiers:
+            if isinstance(arg, arg_type):
+                return arg.get_payload()
+
+        return MISSING
+
+
+class Modifier(metaclass=_CommandTypeMeta):
+    """
+    Modifies a CommandArg by being added to
+    :class:`pincer.commands.arg_types.CommandArg`'s args.
+    """
+
+
+class Description(Modifier):
+    """
+    Represents the description application command option type
+
+    .. code-block:: python3
+
+        # Creates an int argument with the description "example description"
+        CommandArg[
+            int,
+            Description["example description"]
+        ]
+
+    Parameters
+    ----------
+    desc : str
+        The description for the command.
+    """
+
+    def __init__(self, desc) -> None:
+        self.desc = str(desc)
+
+    def get_payload(self) -> str:
+        return self.desc
+
+
+class Choice(Modifier):
+    """
+    Represents an application command choice
+
+    .. code-block:: python3
+
+        Choices[
+            Choice["First Number", 10],
+            Choice["Second Number", 20]
+        ]
+
+    Parameters
+    ----------
+    name : str
+        The name of the choice
+    value : Union[int, str, float]
+        The value of the choice
+    """
+
+    def __init__(self, name, value) -> None:
+        self.name = name
+        self.value = value
+
+
+class Choices(Modifier):
+    """
+    Represents the choice application command option type
+
+    .. code-block:: python3
+
+        CommandArg[
+            int,
+            Choices[
+                Choice["First Number", 10],
+                20,
+                50
+            ]
+        ]
+
+    Parameters
+    ----------
+    *choices : Union[:class:`pincer.commands.arg_types.Choice`, str, int, float]
+        A choice. If the type is not :class:`pincer.commands.arg_types.Choice`,
+        the same value will be used for the choice name and value.
+    """
+
+    def __init__(self, *choices) -> None:
```

Comment:
Would remove `-> None` and rather typehint `*choices` if possible (same for other cases)